README.md
<h1 align="center">Node Binance Trader (NBT)</h1>

<h4 align="center">NBT is a Cryptocurrency Trading Strategy & Portfolio Management Development Framework for <a href='https://www.binance.com/en-AU/register?ref=141340247' target="_new">Binance</a>.</h4>

## About this fork

This is a fork of the source NBT repo, however the focus of my changes are on the trader.ts and associated components. This is the script that receives the buy/sell signals from the NBT Hub (BitcoinVsAltcoins.com) and executes them on Binance. I have not modified the server.

**Important Note:** Many of the new features (e.g. transaction and balance history) are held in memory while the trader is running. The trader will attempt to write these to a configured Postgres database so that important information can be reloaded if the trader restarts. Using the database is optional, but without it you will lose the history every time the trader restarts. If you choose to deploy to Heroku then the database will automatically be set up.

The new features that I have added to the trader include:
* ***CONFIG:* Quantity as Fraction**
  * If enabled, this will interpret the "Quantity to spend per trade" that you have configured on the NBT Hub as a fraction of your total balance in [Binance](https://www.binance.com/en-AU/register?ref=141340247). E.g. if you say 0.1 BTC it will actually spend 10% of your balance on each trade. However, it will only use the balance from a single wallet, it will not add the two together. This will apply to both SHORT and LONG trades, even though SHORT will not use the actual balance. Both trades will always be calculated relative the **Primary Wallet** (see below), regardless of which wallet is actually chosen to execute the trade.
  * As the funds can be locked up in active trades, the trader will calculate an estimated total balance rather than just using the current free balance. This may not always be exactly right because it uses the original price when the trades were opened, not the current market price.
  * To improve performance the trader will cache the balance of each wallet for up to 24 hours. So if you have manually moved funds in Binance it is best to restart the trader after so that it picks up the new balances.
  * It is never a good idea to use 100% of your balance for a single trade. The rule of thumb is about 10%, but you can choose higher or lower depending on your preferred level of risk. The calculated PnL that is displayed on the NBT Hub does not take into consideration the trade quantities. So a strategy with a reported PnL of 500% would only increase your balance by a maximum of 50% if you set the trade quantity to 10%.
* ***CONFIG:* Funding / Auto-Balancing for LONG Trades**
  * Several options are now available for choosing where the funds come from for LONG trades. Please make sure you fully understand the implications of these before using them, as they can provide greater gains but can also cause greater losses. I encourage you to take a look at the code and understand the decisions that are being made by each model.
    * **(Default)**: This will use the available funds in margin or spot wallets, and will stop opening new trades when funds run out (low risk).
    * **Borrow Minimum**: This will use the available funds in margin or spot wallets, and will then start borrowing funds in margin to execute new trades (medium risk).
    * **Borrow All**: This will always borrow funds in margin to execute new trades regardless of available funds (high risk).
    * **Sell All**: This will use the available funds in margin or spot wallets, then re-sell a small portion of every active trade in order to free up funds for the new trades. The aim is to get all active trades to have the same investment. E.g. 4 active trades will use 25% of your balance each, if a 5th trade opens then it will sell a portion from the previous 4 so that all 5 trades are using 20% of your balance. However, once trades start to close it will not re-buy the remaining open trades, so this can result in some variation over time.
    * **Sell Largest**: This will use the available funds in margin or spot wallets, then re-sell half of the largest trade. E.g. (assuming you set the fraction quantity to 1) 1st trade will use 100% of your balance, the 2nd trade will sell 50% of the first trade and use that, the 3rd trade will sell 50% of the first trade so now you have 1st @ 25%, 2nd @ 50%, 3rd @ 25% of your total balance. The difference between this and the "All" model is that it only re-sells from a single existing trade whenever a new one comes in. This means that sometimes the trade sizes will not be equal. This also will not re-buy for remaining open trades.
    * **Sell Largest PnL**: This works much the same as **Sell Largest** above, but instead of taking the trade with the largest cost it will try to find a large trade with the best change in price (profit or loss) since it was opened. This gives some of the losing trades more time to recover before selling. It will only consider trades whose cost is above the average, this means only half the active trades can be considered, but often it is less depending on how many larger vs. smaller trades there are. The disadvantage over the simple **Sell Largest** option is that there will be a slighly longer delay, as it needs to retrieve the current prices before making a decision.
  * If a trade has been manually stopped, it will not be touched by the **Sell All**, **Sell Largest**, or **Sell Largest PnL** options.
  * Each model will first attempt to use the "Quantity to spend per trade" from the NBT Hub (either as an absolute or a fraction), this will represent the largest possible trade value. Only if there are insufficient funds, then it will apply one of the options or reduce the trade quantity.
  * Only LONG trades consume your available balance (all SHORT trades are funded through borrowing up to your maximum limit).
* ***CONFIG:* Prevent Losses for Auto-Balancing**
  * This can only be enabled if using one of the **Sell All**, **Sell Largest**, or **Sell Largest PnL** funding models above. Before choosing which LONG trade(s) to sell it will check the current price on Binance and try to determine which trades would make a loss if they were to be sold now. Any losing trades will be excluded from the auto-balancing logic. This means that if all open trades are losing then it will not sell anything and therefore may not be able to open the new trade, or it may only be able to use the free balance. It could also repeatedly sell the same trade on multiple open signals if that is the only trade currently in profit.
  * Therefore the benefit is that this will minimise losses from auto-balancing, but the downside is that it will limit the opportunities for new trades to open during peak periods, and may result in disproportionate trade sizes.
  * The profit and loss calculation is done with the current ticker price, not the actual market bid price, plus there is always risk of slippage. So there may still be some minor losses due to variation in the actual sell price.
* ***CONFIG:* Repay Loan Interest**
  * Before repaying a margin loan the trader will check your current BNB margin balance info to see how much interest has been accumulated. It will then use your BNB to repay that interest first before repaying the principal loan. Obviously you will need to periodically top up your BNB balance to cover the interest repayments.
  * For this to work properly you need to enable the "Using BNB For Interest" option in your margin wallet in Binance. This means that regardless of what coin is borrowed, it will always accumulate the interest as BNB.
  * If there are multiple open trades with borrowed funds, when the first one closes it will repay all the interest accumulated from all loans, so then the next trade to close will only have to repay anything new.
  * Be aware that because the interest is accrued and repaid in BNB only, it will affect the calculated PnL that is displayed in the **Web Diagnostics**. Interest accrued from other coins will not be reflected in their own PnL, and if you have a strategy that accumulates BNB then it will have a lower PnL as a result of all interest.
* ***CONFIG:* Primary Wallet**
  * You can choose whether to use the spot or margin wallet as your primary wallet for LONG trades. It will also use the total balance from this primary wallet to calculate the size of SHORT trades if you are using **Quantity as Fraction**.
  * The default is margin.
* ***CONFIG:* Wallet Buffer**
  * As slippage, spread, and bad trades are difficult to predict, it is good to keep some additional funds in your wallet to cover these costs. You can specify a buffer amount as a fraction of your wallet which will not be used for opening LONG trades.
  * The default is 0.02 which is 2%.
* ***CONFIG:* Maximum Count of Trades**
  * You can set maximum counts for SHORT and for LONG trades to limit how many can be active at one time. This can be used to limit your exposure to risk when borrowing, or to limit the number of times a LONG trade gets rebalanced. If a trade is stopped it will not count towards the limit.
  * The defaults are zero, which is unlimited.
* ***CONFIG:* Disable SHORT Trades**
  * Because SHORT trades will always borrow the full amount of the trade, you can choose to ignore SHORT trades to prevent borrowing.
  * You can still choose to allow borrowing for LONG trades using one of the borrow funding options above.
* ***CONFIG:* Disable Margin Trades**
  * This will prevent any trades from executing on your margin wallet. All LONG trades will then execute on spot, and SHORT trades will be disabled as a result.
  * Even if a LONG trade is executed on your margin wallet it will not borrow any funds by default. You would have to choose one of the borrow funding options above.
* ***CONFIG:* Disable Coins**
  * You can provide a comma delimited list of coins that you want to ignore trade signals for (e.g. DOGE).
* ***CONFIG:* Strategy Loss Limit**
  * You can set a limit on the number of sequential losses for a strategy. If this many losing trades occur for a strategy then that strategy will be stopped. The trader will ignore all open signals for that strategy, and will only process close signals if the price is better (i.e. higher than the open price for LONG trades and lower for SHORT trades). You can still manually close the trades regardless of the price.
  * Stopping and resuming the strategy via the **Web Diagnostics** will reset the count of losing trades, toggling the trade option via the NBT Hub will also reset the count.
  * The default is zero, which is unlimited.
* ***CONFIG:* Strategy Loss Limit Threshold**
  * Used in conjunction with the **Strategy Loss Limit** this allows early limiting of open trades before stopping the strategy completely. If the threshold of losing trades is reached, then new open signals will only be allowed if the count of losses plus the count of open trades is below the loss limit. Essentially this tries to limit the number of open trades so that if they all closed for a loss then it would not exceed the loss limit. But because this only activates when the threshold of losing trades is hit, it is still possible to exceed the loss limit if many open signals came in before the threshold was reached.
  * The threshold is set as a decimal fraction of the loss limit. For example, if the loss limit was 10 and the threshold was 0.4 then it would only activate when there were 4 losses in a row. Before 4 losses, trading would be normal with no limit on open trades. But once it hits 4 losses then the trader would only allow 6 more trades. If there were 4 losses in a row but already 3 open trades, only 3 more open signals would be allowed through for that strategy, to try to limit the potential run of losses to 10. If any one of those trades closed for profit then everything would reset and all new open signals would be accepted.
  * The default is 0.5, set it to 1 to disable, or 0 to essentially have a fixed trade count per strategy.
* ***CONFIG:* BNB Free Threshold**
  * After a trade is closed it will check your remaining BNB balance and generate a warning if it has dropped below this threshold. This helps you to manage your BNB balance so that there are always sufficient funds to cover fees and interest. These warnings will be displayed in the log and also sent using the notification systems. It will check the spot and margin wallets independently.
  * The default is 0.04 BNB, you can set it to 0 if you only want a warning when it is completely empty, or -1 to disable the check.
* ***CONFIG:* BNB Free Top Up Float**
  * You can set a float level for your free BNB balance, then once your BNB balance drops below this level you can use the top up feature on the PnL Web Diagnostics page. This will convert the chosen coin to BNB to increase the free balance to the float level. There are separate buttons for each wallet.
  * It is better to use this option rather than converting manually through Binance so that the trader can adjust the PnL for the change in balance. It does this by offsetting the fees in the balance history.
  * BNB top ups will not appear in the transactions list.
  * The default is 0.1 BNB, you can set it to 0 if you do not want to allow the top up feature.
* ***CONFIG:* BNB Free Auto Top Up**
  * If you do not want to manually top up the BNB you can set a chosen coin (e.g. BTC) to always use for topping up. If set, this will automatically attempt to purchase BNB using the specified coin and restore the balance to the **Top Up Float** as soon as the balance drops below the **Threshold**.
  * If it succeeds or fails the result will be logged and a notification will be sent.
  * The default is blank, which is disabled.
* ***CONFIG:* Estimated Taker Fee**
  * In a typical setup, fees are charged in BNB, therefore they do not affect the balance of the coin used for funding the strategy. So in order to make the calculated PnL more accurate, the estimated fees are calculated based on the spot wallet Taker Fee percentage. It does not currently calculate the interest charged on margin lending.
  * Unlike other settings, this is represented as a percentage and not a decimal fraction. This means you can just copy and paste the fee directly from Binance.
  * The default is 0.075(%). If you have a higher VIP level or rebate on Binance that entitles you to lower fees, you can modify the Taker Fee percentage within the trader to match.
* ***CONFIG:* Minimum Trade Cost Buffer**
  * Sometimes it is possible to get a MIN_NOTIONAL error from Binance if the trade cost is too small. Even though the trader attempts to round up to the minimum cost, slippage in the price can still cause it to fail. So a buffer is used to increase the minimum cost to attempt to avoid this error.
  * The default is 0.2, which is 2% of the minimum cost as defined by Binance.
* ***CONFIG:* Virtual Wallet Funds**
  * You can set a default balance for virtual trades, this allows you to simulate some of the auto-balancing or funding models above. The value represents roughly the equivalent BTC amount. For example, if you set the funds to 1 but you are trading in USDT, it will use the minimum purchase volumes to estimate a 'similar' USDT value of 1 BTC as the starting balance. This is not current market price, it is just a pre-determined scale set by Binance.
  * The default is 0.1 BTC which (at the time of writing) converts to 10,000 USDT.
* **Alternate Wallet Fall Back**
  * If you do not have sufficient funds in your primary wallet (e.g. margin) to make LONG trades, it will automatically try to make the trade from your other wallet (e.g. spot).
* **Database Backup**
  * If configured, it will save the current state of strategies, open trades, virtual balances, and balance history to a Postgres database. It will also save internal logs and transaction history, but with a limit on the total number of records. If the trader restarts it will first attempt to load the previous state of the strategies, open trades, virtual balances, and balance history from the database. It will still read the strategies and trades from the NBT Hub and compare them to the reloaded data, if there are any discrepancies then it will report this in the logs.
  * When deployed on Heroku it will automatically provision a free dev database, which has a limit of 10,000 records. It will automatically delete some of the oldest logs or transactions if the total number of records reaches this limit.
  * If you are concerned that the data in the database is out of sync with the NBT Hub then you can just reset the database (there is a button in Heroku to do this). This will then load the strategies and open trades from the NBT Hub and clear other history. If rebalancing has occurred on open trades it will make its best guess as to what these should be based on the current balances in Binance.
* **Web Diagnostics**
  * You can connect to the trader webserver to view the internal information that is being tracked (e.g. http://localhost:8003/log). The following commands are available:
    * **/log** - Internal log currently held in memory (newest entries at the top).
    * **/log?db=1** - Internal log loaded from the database (newest entries at the top).
    * **/pnl** - Calculated rate of return and history of open and close balances (best estimation based on available data).
    * **/pnl?reset=** - Specify a coin and trading type to clear the Balance History and PnL for that coin and type (e.g. ?reset=BTC:real).
    * **/pnl?topup=** - Specify a coin and wallet to buy BNB to top up the float (e.g. ?topup=BTC:spot).
    * **/strategies** - Configured strategies.
    * **/strategies?stop=** - Specify a strategy ID to shut down trading, this will still allow open trades to close automatically but only when the signal price is a profit.
    * **/strategies?start=** - Specify a strategy ID to resume normal trading.
    * **/strategies?public** - List of all strategies broadcasting public signals since the trader started.
    * **/trades** - Current open trades list.
    * **/trades?hodl=** - Specify a trade ID to HODL, this will only accept a close signal when the price will result in a profit.
    * **/trades?release=** - Specify a trade ID to release the HODL status, this will accept the next close signal from the NBT Hub.
    * **/trades?stop=** - Specify a trade ID to stop trading, this will keep the trade open but ignore close signals from the NBT Hub.
    * **/trades?start=** - Specify a trade ID to resume trading, this will accept the next close signal from the NBT Hub.
    * **/trades?close=** - Specify a trade ID to attempt to close the trade on Binance. You may need to use this if the trader gets out of sync with the NBT Hub.
    * **/trades?delete=** - Specify a trade ID to delete the trade without closing it. This will not remove the trade from the BVA Hub, so you may have to do that separately. It also will not modify the virtual balances nor close the trade on Binance, so this will probably throw out the PnL calculations as it will seem like there was a significant change in the closing balance. You may want to use one of the 'reset' commands to start fresh.
    * **/trans** - Log of actual buy, sell, borrow, repay transactions held in memory (newest entries at the top).
    * **/trans?db=1** - Log of actual buy, sell, borrow, repay transactions loaded from the database (newest entries at the top).
    * **/virtual** - Views the current virtual balances.
    * **/virtual?reset=true** - Clears and reloads the virtual balances and virtual PnL. You can also pass a number on the reset and it will change the default value for **Virtual Wallet Funds** (e.g. ?reset=100).
    * **/graph.html?summary=** - Specify coin and trading type to graph the last 7 days worth of transactions (e.g. ?summary=BTC:real).
  * You can also configure a **Web Password** in the environment variables to restrict access to these commands (e.g. http://localhost:8003/log?mypassword).
* **HODL Open Trades**
  * Using the feature in the **Web Diagnostics** above you can select individual trades to enable HODL (Hold On for Dear Life). Unlike a stopped trade, a HODL trade can still close automatically from a signal but only if the price will result in a profit. HODL can be set on individual trades without having to stop the strategy.
  * By default a HODL trade will not be used for auto-balancing, but if you have enabled the **Prevent Losses for Auto-Balancing** option then a HODL trade may be used once the price is in profit.
* ***CONFIG:* Automatically Close HODL Trades**
  * If enabled, the trader will periodically check the current price of any trades that are set to HODL, or any trades that belong to a strategy that has been stopped (shut down), and will automatically close any trades that are expected to make a profit. The trader does not do a trailing stop loss, so it will close the trade the first time it detects a profit. This will likely mean that the profit will be minimal, and may even make a small loss if slippage occurs. However, the benefit is that it will close the trades at the earliest opportunity rather than waiting for the original strategy to send another close signal.
  * By default this option is disabled. When enabled it will check the latest prices as part of the background process, which is currently on a five minute interval.
* **Individual Tracking of Real / Virtual Trades**
  * In the original trader if you started a strategy in virtual trading and switched to real trading, or vice versa, it would attempt to close trades based on the current status of the strategy, rather than how the trade was originally opened. This means it could try to close a trade on Binance that was originally opened virtually, or never close the open trade on Binance because you've now switched the strategy to virtual. Now, if the trade opened on Binance it will close on Binance even if the strategy has been switched to virtual. If you don't want this to happen, make sure you close or stop the open trades before switching modes.

CHANGELOG.md
## [1.2.3] (2022-04-05)

### Features

* **trader:** added automatic top up for BNB.

### Bug Fixes

* **trader:** added non-discounted fees for adjusting quantity.
* **trader:** added validation for NODE_ENV.

## [1.2.2] (2022-02-18)

### Features

* **trader:** validate if loaded SHORT trades can exist on margin.
* **trader:** check and adjust quantity to match filled buy order.
* **trader:** added option to auto close HODL trades.

## [1.2.1] (2021-10-20)

### Bug Fixes

* **trader:** correctly identifies cross margin pairs.
* **trader:** shut down with error exit code.
* **trader:** prevent search engine indexing.

## [1.2.0] (2021-10-02)

### Bug Fixes

* **trader:** auto-balancing will not occur if rounding is more than 50%, or remaining quantity is too small.
* **trader:** stopped repeating BNB free threshold message.
* **trader:** limited decimal places for numbers in the web UI.
* **trader:** does not generate silly messages if not needed.
* **trader:** changed auto-balancing warning to 5% to reduce log spam.
* **trader:** stop small trades from blocking auto-balancing.
* **trader:** clear loss trade run when manually resuming a strategy.
* **trader:** correct source type for manual close from NBT Hub.

### Features

* **trader:** added estimation of taker fees for better PnL calculation.
* **trader:** added upgrade path for saved data.
* **trader:** added buffer for minimum trade cost.
* **trader:** compare open trades to the source strategy.
* **trader:** added notifier level configuration.
* **trader:** added BNB top up feature and float, including buttons on PnL page.
* **trader:** added Shut Down, Resume, and BVA buttons to Strategies page.
* **trader:** added Reset buttons to Virtual Balances and PnL pages.
* **trader:** decreased wallet buffer to 2%.
* **trader:** tracking public strategies.
* **trader:** added transaction summary graphs in the web UI.
* **trader:** web UI facelift.
* **trader:** added Sell Largest PnL funding option.
* **trader:** added option to exclude losing trades from auto-balancing.
* **trader:** added option to shorten notification messages.
* **trader:** ability to HODL open trades.
* **trader:** additional BNB threshold warning at 50%.
* **trader:** added threshold for soft entry of loss limit.

## [1.1.0] (2021-08-09)

### Bug Fixes

* **trader:** report initial connection failure.

### Features

* **trader:** added option to manually close trades from the web diagnostics.
* **trader:** added BNB free threshold check.
* **trader:** added Stop, Resume, Close, and Delete buttons to Open Trades page.

## [1.0.1] (2021-08-03)

### Bug Fixes

* **trader:** Telegram notifications now work, and /info command will work without configuring a valid Chat ID.

## [1.0.0] (2021-08-01)

### Bug Fixes

* **trader:** margin borrow and repay functions now work.
* **trader:** identify when a trade request was not fulfilled in Binance.
* **trader:** correct calculation of trade quantity to account for LOT_SIZE and precision.
* **trader:** only allow one concurrent open trade per strategy + symbol + position, this matches the NBT Hub rules.
* **trader:** support (as a workaround) for running on Binance Testnet.
* **trader:** use cached market data.
* **trader:** truncate memory logs.

### Features
* **server:** added description and max_concurrent parameters.
* **trader:** configurable logging levels.
* **trader:** a lot more validation.

These are all for **trader**. Refer to the README for more details:
* ***CONFIG:* Quantity as Fraction**
* ***CONFIG:* Funding / Auto-Balancing for LONG Trades**
* ***CONFIG:* Repay Loan Interest**
* ***CONFIG:* Primary Wallet**
* ***CONFIG:* Wallet Buffer**
* ***CONFIG:* Maximum Count of Trades**
* ***CONFIG:* Disable SHORT Trades**
* ***CONFIG:* Disable Margin Trades**
* ***CONFIG:* Disable Coins**
* ***CONFIG:* Strategy Loss Limit**
* ***CONFIG:* Virtual Wallet Funds**
* **Alternate Wallet Fall Back**
* **Database Backup**
* **Web Diagnostics**
* **Individual Tracking of Real / Virtual Trades**
* **Clean Up Stopped Trades**
* **Track Order Price / Cost**
* **Additional Notifications**
* ***CONFIG:* Additional Logging**

## [0.3.1](https://github.com/jsappme/node-binance-trader/compare/0.3.0...0.3.1) (2021-03-08)


### Reverts

* Revert "refactor: reformat code" ([1fd45ae](https://github.com/jsappme/node-binance-trader/commit/1fd45aeb6fd70c37d8d998ca040ab1767ac91e33))

# [0.3.0](https://github.com/jsappme/node-binance-trader/compare/0.2.2...0.3.0) (2021-03-03)


### Bug Fixes

* **backtest:** remove output of pg_connection string ([8d4803c](https://github.com/jsappme/node-binance-trader/commit/8d4803c88db1508d4b6136d34c34ca0302aaf7c1))
* **deps:** unignore npm lockfile ([275b05a](https://github.com/jsappme/node-binance-trader/commit/275b05a5f1de5e0344bf454c059a7cb0df036ebb))
* **docker:** replace yarn with npm ([1b5940a](https://github.com/jsappme/node-binance-trader/commit/1b5940ad3f87ab4ea4ada0809f80fb967cc86c3b))
* **npm:** set private ([3e5bc54](https://github.com/jsappme/node-binance-trader/commit/3e5bc541a28b87fc1c7e15ec127218b26b1947d6))
* **trader:** crash error: can't access `.minQty` of 'undefined' ([0b214a7](https://github.com/jsappme/node-binance-trader/commit/0b214a70ff61abd803bf9cf4ac8c47b335938d00))
* **trader:** don't log unfollowed sell signals ([f7104de](https://github.com/jsappme/node-binance-trader/commit/f7104de828d97341078ee0d0e0e1a6cdcb447d05))


### Features

* add editorconfig ([e050de0](https://github.com/jsappme/node-binance-trader/commit/e050de0ecf3193dbd867bfe9b3c333c348ffc6cb))
* add renovate configuration ([2b6b3df](https://github.com/jsappme/node-binance-trader/commit/2b6b3dff2751bc639725a0bae618e9ab9b14e76f))
* **backtest:** discard abbreviation ([7ba1b8d](https://github.com/jsappme/node-binance-trader/commit/7ba1b8dc06cadedb348244ce860c85f5758998b6))
* **docker:** add Dockerfile ([2b92e07](https://github.com/jsappme/node-binance-trader/commit/2b92e076c99666e15ce170a35c4594b33b2c549d))

src/trader/env.ts
import path from "path"

import dotenv from "dotenv"
import { bool, cleanEnv, port, str, num, email } from "envalid"

import * as packageJson from "../../package.json"
import { LongFundsType, WalletType } from "./types/trader"

export function testOnly(value: string): string | undefined {
    return process.env.NODE_ENV === "test" ? value : undefined
}

export function getDefault(): Readonly<any> {
    dotenv.config({
        path: path.resolve(
            process.cwd(),
            process.env.NODE_ENV === "test" ? ".env.testing" : ".env"
        ),
    })

    return cleanEnv(process.env, {
        // Required configuration
        BINANCE_API_KEY: str({ devDefault: testOnly("BINANCE_API_KEY"), desc: "Your Binance API Key" }),
        BINANCE_SECRET_KEY: str({ devDefault: testOnly("BINANCE_SECRET_KEY"), desc: "Your Binance API Secret" }),
        BVA_API_KEY: str({ devDefault: testOnly("BVA_API_KEY"), desc: "Your BitcoinVsAltcoins.com Account API Key" }),

        // Gmail nofications
        // Make sure you have allowed 'less secure apps' in your Google account: https://myaccount.google.com/lesssecureapps
        IS_NOTIFIER_GMAIL_ENABLED: bool({ default: false, desc: "Selects if Gmail will be used (Enable less secure apps https://myaccount.google.com/lesssecureapps)" }),
        NOTIFIER_GMAIL_ADDRESS: str({ default: "", desc: "Gmail email to get notifications" }),
        NOTIFIER_GMAIL_APP_PASSWORD: str({ default: "", desc: "Gmail password to get notifications" }),

        // To use Telegram, first talk to The Botfather and create a bot on Telegram: https://core.telegram.org/bots#3-how-do-i-create-a-bot
        // The Botfather will give you your token for the HTTP API, and that is what you set to be the TELEGRAM_API_KEY
        // Then talk to the @userinfobot at Telegram, and it will give you your personal receiver ID, and thats what you use for the TELEGRAM_RECEIVER_ID
        IS_NOTIFIER_TELEGRAM_ENABLED: bool({ default: false, desc: "Selects if Telegram will be used" }),
        NOTIFIER_TELEGRAM_API_KEY: str({ default: "", desc: "Telegram Key for your bot (To create one follow https://core.telegram.org/bots#6-botfather)" }),
        NOTIFIER_TELEGRAM_RECEIVER_ID: str({ default: "", desc: "Unique identifier for the target chat (as a number) or username of the target channel (in the format @channelusername)" }),
        
        NOTIFIER_LEVEL: str({ default: "info", choices: [ "info", "success", "warn", "error" ], desc: "Minimum level of notifications that will be sent"}),
        IS_NOTIFIER_SHORT: bool({ default: false, desc: "Selects if the shorter version of the notification messages should be used" }),

        // Additional configuration options for trader features
        IS_BUY_QTY_FRACTION: bool({ default: false, desc: "Uses the 'Quantity to spend per trade' from the NBT Hub as a fraction of your wallet balance (e.g. 0.1 is 10%)" }),
        TRADE_LONG_FUNDS: str({ default: LongFundsType.NONE, choices: Object.values(LongFundsType), desc: "See README for explanation" }), // '', 'borrow min', 'borrow all', 'sell all', 'sell largest', or 'sell largest pnl'
        IS_FUNDS_NO_LOSS: bool({ default: false, desc: "Will not sell trades for a loss when using one of the sell options of TRADE_LONG_FUNDS" }),
        PRIMARY_WALLET: str({ default: WalletType.MARGIN, choices: Object.values(WalletType), desc: "Primary wallet to execute LONG trades ('margin' or 'spot'), it may still swap to the other if there are insufficient funds" }),
        WALLET_BUFFER: num({ default: 0.02, desc: "Decimal fraction of the total balance of each wallet that should be reserved for slippage, spread, and bad short trades (e.g. 0.02 is 2%)" }),
        MAX_SHORT_TRADES: num({ default: 0, desc: "Maximum number of SHORT trades that can be open concurrently (i.e. limit your borrowing), zero is no limit" }),
        MAX_LONG_TRADES: num({ default: 0, desc: "Maximum number of LONG trades that can be open concurrently (i.e. limit borrowing or rebalancing), zero is no limit" }),
        EXCLUDE_COINS: str({ default: "", desc: "Comma delimited list of coins to exclude from trading (e.g. DOGE)" }),
        STRATEGY_LOSS_LIMIT: num({ default: 0, desc: "Number of sequential losses before a strategy is stopped" }),
        STRATEGY_LIMIT_THRESHOLD: num({ default: 0.5, desc: "Decimal fraction of the STRATEGY_LOSS_LIMIT to determine when to start limiting open trades" }),
        IS_AUTO_CLOSE_ENABLED: bool({ default: false, desc: "Used to automatically close HODL trades or any trades for stopped strategies" }),
        IS_TRADE_SHORT_ENABLED: bool({ default: true, desc: "SHORT trades will always borrow the full funds in margin to execute, disable if you don't want this" }),
        IS_TRADE_MARGIN_ENABLED: bool({ default: true, desc: "Used to disable use of margin wallet trading for both LONG and SHORT trades" }),
        IS_PAY_INTEREST_ENABLED: bool({ default: true, desc: "Automatically repays all BNB interest before repaying margin loans" }),
        BNB_FREE_THRESHOLD: num({ default: 0.04, desc: "Creates a warning if your available BNB balance is below the threshold (too low for fees and interest)"}),
        BNB_FREE_FLOAT: num({ default: 0.1, desc: "The BNB top up option will buy enough BNB to return your free balance to this level"}),
        BNB_AUTO_TOP_UP: str({ default: "", desc: "Quote currency (e.g. BTC) used to automatically top up BNB to the float once the available balance is below the threshold" }),
        TAKER_FEE_PERCENT: num({ default: 0.075, desc: "The discounted spot trading Taker Fee as quoted on Binance when using BNB"}),
        TAKER_FEE_PERCENT_OTHER: num({ default: 0.1, desc: "The standard spot trading Taker Fee as quoted on Binance when not using BNB"}),
        MIN_COST_BUFFER: num({ default: 0.2, desc: "Decimal fraction to increase the minimum trade cost to avoid MIN_NOTIONAL errors"}),
        VIRTUAL_WALLET_FUNDS: num({ default: 0.1, desc: "The (roughly) equivalent BTC value used as the default starting balance for all virtual wallets" }),
        WEB_PASSWORD: str({ default: "", desc: "Password to restrict access to the internal diagnostics webserver" }),

        // Internal configuration options that can be changed but shouldn't need to be
        TRADER_PORT: port({ default: 8003, desc: "The port to trader webserver runs" }),
        VERSION: str({ default: packageJson.version }),
        LOG_LEVEL: str({ default: "info", choices: [ "silly", "debug", "info", "warn", "error" ]}),
        MAX_LOG_LENGTH: num({ default: 1000, desc: "Maximum number of entries for in-memory logging and transactions" }),
        MAX_DATABASE_ROWS: num({ default: 10000, desc: "Maximum number of records that can be written to the database" }),
        MAX_WEB_COLOURS: num({ default: 10, desc: "Maximum number unique colours used to differentiate string data in the web UI" }),
        MAX_WEB_PRECISION: num({ default: 8, desc: "Maximum number decimal places to display for Big Numbers in the web UI and notification messages" }),
        MAX_WEB_GRAPH_DAYS: num({ default: 7, desc: "Maximum number days to display on the transaction summary graphs in the web UI" }),
        REFERENCE_SYMBOL: str({ default: "BNBBTC", desc: "Uses this market data to calculate wallet funds for other coins" }),
        BALANCE_SYNC_DELAY: num({ default: 1500, desc: "Number of milliseconds to wait after making a trade before fetching current balances" }),
        BACKGROUND_INTERVAL: num({ default: 300000, desc: "Number of milliseconds between running the background process" }),
    })
}

export function setDefault(data?: Readonly<any>): void {
    env = data || getDefault()
}

let env: Readonly<any> = {}
setDefault()

export default (): Readonly<any> => env
